Make collections rendering multiple templates automatically cacheable.

kaspth opened the pull request:
Followup to #18948.

Before this if you had a collection rendering more than one template, say the templates like `_notification` and `_amazing_notification`, we couldn't automatically cache them.

Now we should be able to do that.

So people who helped out on the first one, @dhh, @rafaelfranca, @jeremy and anybody else interested, pipe in :smile:

(I've also changed `automatic_cache_eligible?` to no longer check if the cache key is callable. We were already making that work elsewhere.)

dhh commented:
Rocking! I'll let @jeremy chime in on anything on the implementation.

rafaelfranca commented on actionpack/test/controller/caching_test.rb:
```diff
@@ -406,4 +419,25 @@ def test_caching_works_with_beginning_comment
+  def test_caching_works_with_collection_rendering_multiple_templates
+    get :index_with_multiple_templates
+
+    ActionView::PartialRenderer.expects(:collection_without_template).never
```
Reviewing this test I don't know what this means. Why I should expect this method to never be called?

kaspth replied:
I agree, it's terrible.

Internally `render_collection` calls `collection_without_template` to actually render the templates. I couldn't think of a better way to test that no new templates were rendered.

(I'm also a bit wary of mocks and stubs as it feels like there's a greater chance of false positives there to me.)

rafaelfranca replied:
Ah, so my font of confusion is the name of the method. For me `collection_without_template` is not say what it means, that is render the template for the collection. Maybe we should rename it to `render_collection_without_template`?

kaspth replied:
It's definitely confusing. Especially because it renders a collection with many templates (I guess if you think of it as rendering a collection without a fixed template it makes sense).

--
Kasper

> Den 01/04/2015 kl. 05.35 skrev Rafael Mendonça França <notifications@github.com>:
> 
> In actionpack/test/controller/caching_test.rb:
> 
> > @@ -406,4 +419,25 @@ def test_caching_works_with_beginning_comment
> >      ActionView::PartialRenderer.expects(:collection_with_template).never
> >      get :index_with_comment
> >    end
> > +
> > +  def test_caching_works_with_collection_rendering_multiple_templates
> > +    get :index_with_multiple_templates
> > +
> > +    ActionView::PartialRenderer.expects(:collection_without_template).never
> Ah, so my font of confusion is the name of the method. For me collection_without_template is not say what it means, that is render the template for the collection. Maybe we should rename it to render_collection_without_template?
> 
> —
> Reply to this email directly or view it on GitHub.
>

rafaelfranca commented on actionview/lib/action_view/renderer/partial_renderer.rb:
```diff
@@ -439,6 +439,11 @@ def collection_with_template
         content = template.render(view, locals)
         content = layout.render(view, locals) { content } if layout
+
+        if template.eligible_for_collection_caching?(as: as)
```
We changed `collecton_with_template` so maybe we need a test?

kaspth replied:
I'm not sure we need to cache fixed template renders here actually.
Thinking about this some more I'm scared we might be doing multiple writes, which is... not good :grin:

--
Kasper

> Den 01/04/2015 kl. 05.36 skrev Rafael Mendonça França <notifications@github.com>:
> 
> In actionview/lib/action_view/renderer/partial_renderer.rb:
> 
> > @@ -439,6 +439,11 @@ def collection_with_template
> >  
> >          content = template.render(view, locals)
> >          content = layout.render(view, locals) { content } if layout
> > +
> > +        if template.eligible_for_collection_caching?(as: as)
> We changed collecton_with_template so maybe we need a test?
> 
> —
> Reply to this email directly or view it on GitHub.
>

mtomov commented:
Hello, and thank you for this amazing piece of smart logic. I am surprised to see that you were able to extend the previous implementation to work with different partials now in so very few lines, and without deleting code.

What is the hold-up on merging this onto master?


And, I also agree that the `collection_without_template` has to have its name changed, as when I was looking at the code, it was very difficult to understand that there is actually a template, but it is dynamically invoked. 

Thank you again, brilliant stuff.

kaspth commented:
Thanks for the kind words :grin:

Partly because I didn't want to just throw more lines into the collection rendering (they feel too tackled on for my taste).

Hm, I also remember it as being that cache might not be invalidated, but seems my last comment was me being concerned that it wrote to the cache more than once.

kaspth committed:
Make automatic collection caching work with several templates.

To make automatic collection caching work when a collection renders more than
one template, we need to do several things:

* Eager load templates to get correct `virtual_path` from when generating
  `expanded_cache_key`s.

* When some partials are read from cache, the collection is mutated to remove
  the cached templates. In doing so, the `collection_data` is out of sync,
  its indexes no longer tracking the right objects.

  Since we know the indexes when we shim the collection, we can capture those
  and yield them when rendering. We shift the `partial_iteration` indexes to
  match what they would have been had the collection never been cached.

Note that I think this implementation is a hot mess and, I wager, a nightmare
to maintain. Would love to find a way to make this easier to understand.

kaspth commented:
@dhh @rafaelfranca So... I've had one hell of a weekend down the rabbit hole here. Lots of stumbling, but I think I got collections rendering multiple templates to be cached and `read_multi`'ed. But it wasn't without cost. I'm not sure if we want this implementation.

We could save some lines by not supporting the `PartialIteration` stuff and just reassigning to `@collection_data`.

kaspth commented on actionview/lib/action_view/renderer/partial_renderer.rb:
```diff
@@ -394,9 +428,11 @@ def setup(context, options, block)
         @template_keys = retrieve_template_keys
       else
-        paths.map! { |path| retrieve_variable(path, as).unshift(path) }
+        @collection_data.map! { |path| retrieve_variable(path, as).unshift(path) }
```
Notice how `@collection_data` is assigned to `paths` on line 415. Found that confusing as hell, so I'd rather use the ivar.

kaspth commented on actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb:
```diff
@@ -12,16 +12,27 @@ module CollectionCaching # :nodoc:
+            captured_render_indexes << index
+            collection_to_render << object
+          end
+        end.values
```
Dang left over. Will remove `values` later.

kaspth commented:
Ah, great. Now we're getting hit by `view_cache_dependencies` being a helper method. Remember that being fun the first time around as well :sweat_smile:

dhh commented:
When you're talking about cost, are you referring to the complexity of the implementation or the runtime performance?

kaspth commented:
> When you're talking about cost, are you referring to the complexity of the implementation or the runtime performance?

10-month respond time!… I can't remember 😅

---------------------------------------

@dhh you had solved this in Basecamp with an intermediate template, right? Maybe that solution could serve as inspiration here, somehow?

kaspth commented:
Took it off 5.1 for now — don't think I have the time to sort out at the moment.
